Setup mTLS gateway and access exposed workload with certificate tutorials (#15505)
kyma-project/kyma: docs/03-tutorials/00-security/sec-02-setup-mtls-gateway.md

---
title: Set up a mutual TLS Gateway 
---

This tutorial shows how to set up a mutual TLS Gateway and configure authentication based on certificate details.

## Prerequisites

> To follow this tutorial, use Kyma 2.6 or higher.

Reviewer: is that the case?

Reviewer 2: it's not the case. Kyma ma version is not related

Before you start, Set up a [`custom-domain`](../00-api-exposure/apix-02-setup-custom-domain-for-workload.md) and prepare a certificate to expose a workload.

## Steps

1. Export the following values as environment variables:

   ```bash
   export DOMAIN_TO_EXPOSE_WORKLOADS={DOMAIN_NAME} 
   ```
   >**NOTE:** `DOMAIN_NAME` is the domain that you own, for example, api.mydomain.com

2. Create a Namespace and export its value as an environment variable. Run:
   >**NOTE:** Skip this step if you already have a Namespace

      ```bash
   export NAMESPACE={NAMESPACE_NAME}
   kubectl create ns $NAMESPACE
   kubectl label namespace $NAMESPACE istio-injection=enabled --overwrite
   ```

3. Create Client Root CA and CLient certificate signed by Client Root CA. Export the following values as environment variables and run the command provided:  
      ```bash

>     export CLIENT_ROOT_CA_KEY_FILE=client-root-ca.key

Reviewer: which of these exports can be re-used in expose and secure workload? i would like to avoid duplication

       namespace: istio-system
   EOF
   ```
7. Manage Client TLS
   Create Client Root CA and CLient certificate signed by Client Root CA
   ```bash
   # Create root ca key and cert (valid for 5 years) - will be used for validation
   openssl req -x509 -sha256 -nodes -days 1825 -newkey rsa:2048 -subj '/O=example Inc./CN=ClientRootCA' -keyout ${CLIENT_ROOT_CA_KEY_FILE} -out ${CLIENT_ROOT_CA_CRT_FILE}
   # Create new key and csr for client cert
   openssl req -out ${CLIENT_CERT_CSR_FILE} -newkey rsa:2048 -nodes -keyout ${CLIENT_CERT_KEY_FILE} -subj "/CN=client.example.com/O=example"
   # Sign client cert with CA cert
   openssl x509 -req -days 365 -CA ${CLIENT_ROOT_CA_CRT_FILE} -CAkey ${CLIENT_ROOT_CA_KEY_FILE} -set_serial 0 -in ${CLIENT_CERT_CSR_FILE} -out ${CLIENT_CERT_CRT_FILE}
   # Create new key and csr for client cert
   openssl req -out ${CLIENT_CERT_CSR_FILE}.2 -newkey rsa:2048 -nodes -keyout ${CLIENT_CERT_KEY_FILE}.2 -subj "/CN=client2.example.com/O=example"
   # Sign client cert with CA cert
   openssl x509 -req -days 365 -CA ${CLIENT_ROOT_CA_CRT_FILE} -CAkey ${CLIENT_ROOT_CA_KEY_FILE} -set_serial 0 -in ${CLIENT_CERT_CSR_FILE}.2 -out ${CLIENT_CERT_CRT_FILE}.2
   
   ```

> 8. Add Client Root CA to cacert bundle secret for mTLS Gateway

Reviewer 3: ```suggestion
4. Add Client Root CA to cacert bundle secret for mTLS Gateway
  Export the following values as environment variables and run the command provided.

   ```bash
   export MTLS_GATEWAY_NAME=mtls-gateway
   ```
```

   ```bash
   # Add CA Cert to kyma-mtls-gateway
   export CLIENT_ROOT_CA_CRT_ENCODED=$(cat ${CLIENT_ROOT_CA_CRT_FILE}| base64)
   cat <<EOF | kubectl apply -f -
   ---
   apiVersion: v1
   kind: Secret
   metadata:

>      name: ${MTLS_CERT_SECRET_NAME}-cacert

Reviewer 3: ```suggestion
     name: ${TLS_SECRET}-cacert
```

     namespace: istio-system
   type: Opaque
   data:
     cacert: ${CLIENT_ROOT_CA_CRT_ENCODED}
   EOF
   ```

9. Create mTLS Gateway (mode: MUTUAL)
   ```bash
   cat <<EOF | kubectl apply -f -
   ---
   apiVersion: networking.istio.io/v1beta1
   kind: Gateway
   metadata:
     name: ${MTLS_GATEWAY_NAME}

>      namespace: ${MTLS_TEST_NAMESPACE}

Reviewer 3: ```suggestion
     namespace: ${NAMESPACE}
```

   spec:
     selector:
       istio: ingressgateway
       app: istio-ingressgateway
     servers:
       - port:
           number: 443
           name: https
           protocol: HTTPS
         tls:
           mode: MUTUAL

>            credentialName: ${MTLS_CERT_SECRET_NAME}

Reviewer 3: ```suggestion
           credentialName: ${TLS_SECRET}
```

           minProtocolVersion: TLSV1_2
           cipherSuites:
           - ECDHE-RSA-CHACHA20-POLY1305
           - ECDHE-RSA-AES256-GCM-SHA384
           - ECDHE-RSA-AES256-SHA
           - ECDHE-RSA-AES128-GCM-SHA256
           - ECDHE-RSA-AES128-SHA
         hosts:

>            - '*.${CUSTOM_DOMAIN}'

Reviewer 3: ```suggestion
           - '*.${DOMAIN_TO_EXPOSE_WORKLOADS}'
```

       - port:
           number: 80
           name: http
           protocol: HTTP
         tls:
           httpsRedirect: true
         hosts:

>            - '*.${CUSTOM_DOMAIN}'

Reviewer 3: ```suggestion
           - '*.${DOMAIN_TO_EXPOSE_WORKLOADS}'
```
